doc/devices/drivers/gpio/gpio.rst

```diff
@@ -0,0 +1,294 @@
+
+**GPIO_PUD_PULL_DOWN**
+    Enable GPIO pin pull-down.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_BIAS_PULL_DOWN`
+
+**GPIO_DS_DFLT_LOW**
+    Default drive strength standard when GPIO pin output is low.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_STRENGTH_DEFAULT`
+
+**GPIO_DS_DFLT_HIGH**
+    Default drive strength when GPIO pin output is high.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_STRENGTH_DEFAULT`
+
+**GPIO_DS_ALT_LOW**
+    Alternative drive strength when GPIO pin output is low.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_STRENGTH_7`
+
+**GPIO_DS_ALT_HIGH**
+    Alternative drive strength when GPIO pin output is high.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_STRENGTH_7`
+
+**GPIO_DS_DISCONNECT_LOW**
+    Disconnect pin when GPIO pin output is low.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_OPEN_SOURCE`
+
+**GPIO_DS_DISCONNECT_HIGH**
+    Disconnect pin when GPIO pin output is high.
+
+    PINCTRL pin configuration property:
+
+    - :c:macro:`PINCTRL_CONFIG_DRIVE_OPEN_DRAIN`
+
+Signalling Configuration
+========================
+
+Pin signalling can be configured by the general GPIO configuration interface.
+
+:c:func:`gpio_pin_configure`
+    Set the configuration of a pin.
+
+A generic set of GPIO port pin signalling properties is defined in
+:file:`include/gpio.h`. A GPIO driver may only support a subset of the
+signalling properties.
+
+**GPIO_INT**
+    GPIO pin to trigger interrupt.
+
+**GPIO_INT_ACTIVE_LOW**
+    GPIO pin trigger on level low or falling edge.
+
+**GPIO_INT_ACTIVE_HIGH**
+    GPIO pin trigger on level high or rising edge.
+
+**GPIO_INT_CLOCK_SYNC**
+    GPIO pin trigger to be synchronized to clock pulses.
+
+**GPIO_INT_DEBOUNCE**
+    Enable GPIO pin trigger debounce.
+
+**GPIO_INT_LEVEL**
+    Do Level trigger.
+
+**GPIO_INT_EDGE**
+    Do Edge trigger.
+
+**GPIO_INT_DOUBLE_EDGE**
+    Interrupt triggers on both rising and falling edge.
+
+
+GPIO device tree support
+************************
+
+Nodes
+=====
+
+GPIO Controller Node
+--------------------
+
+The GPIO device is defined by a GPIO controller node.
+
+Required properties are:
+
+    - gpio-controller: Indicates this device is a GPIO controller
+    - compatible:
+    - label: Should be a name string
+
+
+Interaction with the PINCTRL device driver
+==========================================
+
+To enable a pin controller to report the pins that are associated to a GPIO device
+the gpio-ranges property shall be set in the device tree. Also the label property
+shall be set and used as device name.
+
+The gpio-ranges property is of the form:
+
+.. code-block:: text
+
+    gpio-ranges = <&[pin controller] [pin number in GPIO pin number space]
+                                     [pin number in pin controller number space]
```

Comment:
Can you elaborate on the of opportunity of introducing pin controller space concept in Zephyr?
How will it work and be used ?

Specially, I hardly see how this could be used when some GPIO ports/nodes could be deactivated from one build to another (say if GPIO port C is not used), in that case 16 pins will be removed from the total pin controller number space. Then how to work with that kind of moving references.
On the other hand, if total pin controller number space stays constant even when a GPIO port is disabled from the build, then we'll reserve pin that don't "exist" in HW and then we're wasting footprint.

Reply:
> Can you elaborate on the of opportunity of introducing pin controller space concept in Zephyr?
How will it work and be used ?

The opportunity is more on the GPIO controller side. Having a GPIO port controller with its own pin number space allows to have a generic implementation for all ports. It also allows to code the pin numbers in a way to be able to define sets of GPIO pins for operations on a set of GPIO pins (PIN0 = 0x0001, PIN1=0x0002, ... PIN15 = 0x8000). The mapping between these pin numbers and the ones used by the pin controller is done by the gpio-ranges directive in the device tree.

Using the same coding as for GPIOs for the pin controller pin numbers would not work. A driver implementer would therefor most probably choose to use consecutive pin numbers. You may see that as another opportunity. Driver writers are free to choose the pin numbering scheme that fits best to their driver implementation. Usually a driver provides a header file with defines for the pin numbers. 

> Specially, I hardly see how this could be used when some GPIO ports/nodes could be deactivated from one build to another (say if GPIO port C is not used), in that case 16 pins will be removed from the total pin controller number space. Then how to work with that kind of moving references

The pin controller pin number space is independent from GPIOs activated or not activated. A pin controller controls a set of physical pins - independent whether these pins are used  by a GPIO driver or not. If you de-activate the GPIO port the GPIO driver for the port will not be instantiated but the physical pins are still there and can be controlled by the pin controller.

Another case is if GPIO C is not available in a specific SoC variant. A pin controller driver that covers all SoC variants and must be generic would just treat that as a whole in the pin number  space. There is no guarantee that pin numbers are consecutive. I mentioned that in the documentation.

> On the other hand, if total pin controller number space stays constant even when a GPIO port is disabled from the build, then we'll reserve pin that don't "exist" in HW and then we're wasting footprint.

The pin controller driver holds resources for the information provided in the device tree. If there is no information regarding the non-existing GPIO port pins then there is also no wasted footprint. You have to assure that the non-existing GPIO port is deactivated in the device tree.